packages/backend/src/logging/caslAuditWrapper.ts

```diff
@@ -0,0 +1,201 @@
+import type { Rule as CaslRule } from '@casl/ability/dist/types/Rule';
+import { Abilities, type AnyObject } from '@casl/ability/dist/types/types';
+import { type SessionUser } from '@lightdash/common';
+import {
+    AuditActor,
+    AuditContext,
+    AuditResource,
+    AuditStatusType,
+    createAuditLogEvent,
+    type AuditLogEvent,
+} from './auditLog';
+
+export type AuditLogger = (event: AuditLogEvent) => void;
+
+export type AuditableUser = Pick<
+    SessionUser,
+    | 'userUuid'
+    | 'email'
+    | 'firstName'
+    | 'lastName'
+    | 'organizationUuid'
+    | 'role'
+>;
+
+type AuditHelperArgs = {
+    user: AuditableUser;
+    action: string;
+    subject: AnyObject;
+    ip?: string;
+    userAgent?: string;
+    requestId?: string;
+    ruleConditions?: string;
+};
+
+/**
+ * Creates an audit actor from a user
+ */
+const createActorFromUser = (user: AuditableUser): AuditActor => ({
+    id: user.userUuid,
+    email: user.email || '',
+    firstName: user.firstName || '',
+    lastName: user.lastName || '',
+    organizationUuid: user.organizationUuid || '',
+    organizationRole: user.role || 'unknown',
+    // TODO: Add group memberships
+    groupMemberships: [],
+});
+
+const createResourceFromSubject = (subject: AnyObject): AuditResource => {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    const { __caslSubjectType__, ...rest } = subject;
+
+    // Ensure we have a valid organizationUuid
+    const organizationUuid = (rest.organizationUuid as string) || '';
+
+    return {
+        type: (__caslSubjectType__ as string) || 'unknown',
+        id: (rest.id || rest.uuid || rest.projectUuid || organizationUuid) as
+            | string
+            | undefined,
+        name: (rest.name as string) || undefined,
+        organizationUuid,
+        projectUuid: (rest.projectUuid as string) || undefined,
+    };
+};
+
+const createContextFromArgs = (args: AuditHelperArgs): AuditContext => ({
+    organizationId: args.user.organizationUuid,
+    projectId: args.subject.projectUuid as string | undefined,
+    ip: args.ip,
+    userAgent: args.userAgent,
+    requestId: args.requestId,
+});
+
+// Helper function to extract conditions from a CASL Rule
+const extractRuleConditions = <A extends Abilities, C>(
+    rule: CaslRule<A, C> | null,
+): string | undefined => {
+    if (!rule) return undefined;
+
+    try {
+        // Get conditions directly from Rule object's conditions property
+        const { conditions } = rule;
+
+        if (conditions && typeof conditions === 'object') {
+            return JSON.stringify(conditions);
+        }
+
+        return undefined;
+    } catch (e) {
+        return undefined;
+    }
+};
+
+export class CaslAuditWrapper<T extends Ability> {
+    private wrappedAbility: T;
+
+    private user: AuditableUser;
+
+    private ip?: string;
+
+    private userAgent?: string;
+
+    private requestId?: string;
+
+    private auditLogger: AuditLogger;
+
+    constructor(
```

Comment:
Would be nice to see an example of how this would look like in a service. 
At the moment, the user contains the ability so should these be separate arguments?

Reply:
Adding